ModuleKit/module.swift

```diff
@@ -0,0 +1,287 @@
+//
+//  module.swift
+//  ModuleKit
+//
+//  Created by Serhiy Mytrovtsiy on 09/04/2020.
+//  Copyright © 2020 Serhiy Mytrovtsiy. All rights reserved.
+//
+
+import Cocoa
+import os.log
+import StatsKit
+
+public protocol Module_p {
+    var available: Bool { get }
+    var enabled: Bool { get }
+    
+    var widget: Widget_p? { get }
+    var settings: Settings_p? { get }
+    
+    func load()
+    func terminate()
+}
+
+public struct module_c {
+    public var name: String = ""
+    public var icon: NSImage? = nil
+    
+    var defaultState: Bool = false
+    var defaultWidget: widget_t = .unknown
+    var availableWidgets: [widget_t] = []
+    
+    var widgetsConfig: NSDictionary = NSDictionary()
+    
+    init(in path: String) {
+        let dict: NSDictionary = NSDictionary(contentsOfFile: path)!
+        
+        if let name = dict["Name"] as? String {
+            self.name = name
+        }
+        if let state = dict["State"] as? Bool {
+            self.defaultState = state
+        }
+        
+        if let widgetsDict = dict["Widgets"] as? NSDictionary {
+            self.widgetsConfig = widgetsDict
+            for widgetName in widgetsDict.allKeys {
+                if let widget = widget_t(rawValue: widgetName as! String) {
+                    self.availableWidgets.append(widget)
+                    
+                    let widgetDict = widgetsDict[widgetName as! String] as! NSDictionary
+                    if widgetDict["Default"] as! Bool {
+                        self.defaultWidget = widget
+                    }
+                }
+            }
+        }
+    }
+}
+
+open class Module: Module_p {
+    public var config: module_c
+    
+    public var available: Bool = false
+    public var enabled: Bool = false
+    
+    public var widget: Widget_p? = nil
+    public var settings: Settings_p? = nil
+    
+    private var settingsView: Settings_v? = nil
+    private var popup: NSWindow = NSWindow()
+    
+    private let log: OSLog
+    private var store: UnsafePointer<Store>? = nil
+    private var readers: [Reader_p] = []
+    private var menuBarItem: NSStatusItem = NSStatusBar.system.statusItem(withLength: NSStatusItem.squareLength)
+    private var activeWidget: widget_t {
+        get {
+            let widgetStr = self.store?.pointee.string(key: "\(self.config.name)_widget", defaultValue: self.config.defaultWidget.rawValue)
+            return widget_t.allCases.first{ $0.rawValue == widgetStr } ?? widget_t.unknown
+        }
+        set {}
+    }
+    private var ready: Bool = false
+    private var widgetLoaded: Bool = false
+    
+    public init(store: UnsafePointer<Store>?, popup: NSView?, settings: Settings_v?) {
```

Comment:
Could you explain me, please, what was the reason to inject this dependency as a UnsafePointer? `weak` modifier seems to be enough here

Reply:
Hmm.. Actually you're passing the pointer to the variable that contains a reference to `Store` object, and every `Module` contains sucha indirect reference.

This could be useful if `Store` object is supposed to be replaced with new one on the fly, so that after the change of the value in `store: Store` variable (in AppDelegate.swift) all `Module`s got new `Store` with no additional effort.

IMHO, all these `store: UnsafePointer<Store>` could be replaced with `weak var store: Store?`.

Although I proposed `weak` to keep original retain-free behavior, looks like it is not necessary too, because `store` variable in `AppDelegate` is actually immutable so that a lifetime of `Store` equals to a lifetime of the application.


To bottom line it:
```diff
AppDelegate.swift
- var store: Store = Store()
+ let store: Store = Store()
```
```diff
module.swift 
- private var store: UnsafePointer<Store>
+ private let store: Store

- self.enabled = self.store.pointee.bool(key: ...
+ self.enabled = self.store.bool(key: ...
```
```diff
readers.swift
- public var store: UnsafePointer<Store>? = nil
+ public var store: Store?

- let removableState = store?.pointee.bool(key: ...
+ let removableState = store?.bool(key: ...
```
and so on

Reply:
Thanks for the explanation. Yes, I suppose pointers could be removed at all.